### ISSUE TITLE
Allow using different PK type than ObjectId in mongo driver

### ISSUE THREAD

**Issue by mastermunj** [mikro-orm/mikro-orm#349]
**Describe the bug**
I want to be able to have `_id` as `string` instead of `ObjectId` for Mongo collections. 

I believe the issue is with `denormalizePrimaryKey` which by default converts data into ObjectId at
https://github.com/mikro-orm/mikro-orm/blob/master/lib/platforms/MongoPlatform.ts#L32

I tried searching, however, I couldn't find any workaround. Maybe it's not a popular use case as such.

**Stack trace**
```
Error: Argument passed in must be a single String of 12 bytes or a string of 24 hex characters
    at new ObjectID (/app/node_modules/bson/lib/bson/objectid.js:59:11)
    at MongoPlatform.denormalizePrimaryKey (/app/node_modules/mikro-orm/dist/platforms/MongoPlatform.js:23:16)
    at EntityFactory.create (/app/node_modules/mikro-orm/dist/entity/EntityFactory.js:22:33)
    at EntityManager.create (/app/node_modules/mikro-orm/dist/EntityManager.js:227:40)
    at EntityRepository.create (/app/node_modules/mikro-orm/dist/entity/EntityRepository.js:75:24)
    at UserService.create (/app/dist/modules/user/user.service.js:27:42)
    at UserController.create (/app/dist/modules/user/user.controller.js:22:33)
    at /app/node_modules/@nestjs/core/router/router-execution-context.js:37:29
    at InterceptorsConsumer.intercept (/app/node_modules/@nestjs/core/interceptors/interceptors-consumer.js:10:20)
    at /app/node_modules/@nestjs/core/router/router-execution-context.js:45:60
    at /app/node_modules/@nestjs/core/router/router-proxy.js:8:23
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at next (/app/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/app/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
    at /app/node_modules/express/lib/router/index.js:281:22
```

**Versions**

| Dependency | Version |
| - | - |
| node | 13.8.0 |
| typescript | 3.8.2 |
| mikro-orm | 3.2.1 |
| mongodb | 3.5.1 |

**Comment by B4nan** [comment]
So you would like to have the primary key of entity as `_id: string`, right? That is probably not currently supported and I think it won't be doable without internal changes (so can't think of any workaround now). May I also see the code that produces the error, as well as the entity definition?

**Comment by B4nan** [comment]
But actually I can see a workaround. If you want to have this kind of behaviour for all primary keys, you could extend the existing mongo driver and just override that method so it returns strings. You would need to extend `MongoDriver` and `MongoPlatform`, but the only changes would be in this method plus in the reference to the new platform. 

Not tested, but something like this could work:

```typescript
export class StringMongoPlatform extends MongoPlatform {

  normalizePrimaryKey<T = number | string>(data: Primary<T> | IPrimaryKey | ObjectId): T {
    if (data instanceof ObjectId) {
      return data.toHexString() as unknown as T;
    }

    return data as unknown as T;
  }

  denormalizePrimaryKey(data: number | string): IPrimaryKey {
    return '' + data;
  }

}

export class StringMongoDriver extends MongoDriver {

  protected readonly platform = new StringMongoPlatform();

}

const orm = await MikroORM.init({
  type: 'mongo',
  driver: StringMongoDriver,
});
```

**Comment by mastermunj** [comment]
I tried extending MongoPlatform, however, MongoPlatform is not exported as of now. In fact, neither drivers, connections are exported except the interfaces & default ones. Maybe exporting them will increase extensibility.

Here's my entity definition.

```ts
import {
  Entity,
  IdEntity,
  Property,
  PrimaryKey,
} from 'mikro-orm';


@Entity()
export class User implements IdEntity<User>{

  @PrimaryKey()
  id!: string;

  @Property({ unique: true })
  mobile: string;
}

```

**Comment by B4nan** [comment]
> I tried extending MongoPlatform, however, it is not exported as of now. In fact, neither drivers, connections are exported except the interfaces & default ones. Maybe exporting them will increase extensibility.

That is on purpose, they can't be exported directly from `mikro-orm` package index, as that would require you to install all the driver connectors (for all of them, even if you do not use them).

Just import them from the dist folder: 

```typescript
import { MongoDriver } from 'mikro-orm/dist/drivers/MongoDriver';
import { MongoPlatform } from 'mikro-orm/dist/platforms/MongoPlatform';
```

**Comment by mastermunj** [comment]
Got it. I will give it a try.

Btw, with this approach, if I want to have some collections with ObjectId and some with String Id, will it be possible?

**Comment by B4nan** [comment]
You could probably handle that too, but that would require to override much more methods of the driver (and possibly other classes). And maybe it is not even possible, if I see it correctly, you would have to do changes in the `EntityFactory` which is not extensible as of now. 

But that is pretty weird requirement to be honest :] I guess you are working with same legacy database?

On the other hand, I am fine with supporting this weirdness in the ORM directly, I just do not see it as a priority. I am open for PRs, but it won't be simple change probably.

**Comment by mastermunj** [comment]
I do not work with legacy database as such, however, I do not prefer ObjectId for my entities in general. ObjectId is used only in case I do not need to share the id with others outside the app ever.

**Comment by mastermunj** [comment]
If you can guide me with changes that may be needed, I may be able to raise a PR.

**Comment by B4nan** [comment]
Sounds strange, having this kind of inconsistency in the database on purpose. By sharing the ID you mean you grant someone direct access to database? If not, I do not understand that at all, as how you provide others the ID is not connected to how it is stored. 

> If you can guide me with changes that may be needed, I may be able to raise a PR.
 
Start with setting up the project and running tests, without that such change will be hard to correctly implement. Then followup with designing new mongo entity with this kind of setup, and a new test case where you will use it. There will be multiple things that will need to be changed, you will need to debug the code yourself, if I knew what to change where, I would do it myself :] Probably `EntityFactory` and `EntityHelper` are the places, then possibly `DatabaseDriver`. Just keep in mind that everything that is currently inside the tests needs to work as is. 

https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md

**Comment by B4nan** [comment]
Any update on this? I would love to understand the reason why you need this, as I failed not understand this:

> I do not prefer ObjectId for my entities in general. ObjectId is used only in case I do not need to share the id with others outside the app ever.

Mainly because the ORM will natively handle string PK counterparts automatically for you (`_id` is ObjectId, while `id` is string). 

https://mikro-orm.io/docs/usage-with-mongo/#objectid-and-string-id-duality

**Comment by mastermunj** [comment]
For the following reasons for certain collections/tables, I use custom generated ids.

- When id needs to be communicable, e.g. Customer Id, Order Id and so on.
- Having control of primary key generation makes migrating between databases easier if the need arises. (This should not be problem with ObjectId as such since it can be used as a string in other type of database when needed.)


In the absence of not being able to set the custom primary key, the workaround is to have another id that is custom generated & unique.

**Comment by B4nan** [comment]
Ok so I misunderstood, you want string id you will provide youself (like a uuid), not to use a string version of object id? 

I can totally understand that.

**Comment by B4nan** [comment]
Closing as implemented in v4 via #568. Please subscribe to #527 for updates about it.

**Comment by bogini** [comment]
@B4nan Was this feature subsequently removed? It doesn't look like it's supported in `5.1.5`. Thanks! 🙏
